### ISSUE TITLE
CMake vs plain-header

### ISSUE THREAD

**Issue by havogt** [msimberg/hwh#3]
By quickly browsing the project, I couldn't easily spot why generating the header with CMake is required. Could you add a small section in the README about this design decision?

**Comment by msimberg** [comment]
That's a good question. I'll respond here first and then we can see what makes sense to put in the README.

The primary reason, by far, was ease of development while setting this up. It's of course much easier to change the helper functions than to change the header itself when trying different variations of the wrappers. The alternative way to _generate_ the wrappers would be with macros in the header directly, but I find that this obscures error messages (wrong types to function calls etc.) enough that I did not like this option.

The above argument gets weaker if the functions to generate the wrappers don't change, as one would hope is the case in the future. In this case, copying and pasting function definitions to add new wrappers would not be much more cumbersome than adding a call to one of the CMake generator functions, but generating them is still easier.

Second, it's a bit easier to add the CUDA/HIP variants of the wrappers by generating the wrappers (again, I've excluded macros because of worse error messages). If the header was "hard-coded" there would then be two wrappers to add when adding new wrappers.

Finally, and I'll admit this is not a very strong argument, it's to discourage people from copying and pasting headers into their projects, which I dislike. They can still do it but they'll have to generate the header once first.

TL;DR: From a (CMake) user's perspective a generated or plain header makes no difference and generating is slightly nicer from the (whip) developer's perspective.

All of the above of course ignores non-CMake users which is simply not a concern for us (at the moment?).

Also, generating the header still leaves the path open to create "custom releases" with the header already generated, but I'm avoiding this unless it becomes highly requested.

All of that said, I'd just like to ask you back if generating the header blocks use-cases that you would otherwise have?
